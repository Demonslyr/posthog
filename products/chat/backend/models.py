from typing import TYPE_CHECKING

from django.db import models
from django.db.models import QuerySet

from posthog.models.utils import UUIDModel
from posthog.models.file_system.file_system_mixin import FileSystemSyncMixin
from posthog.models.file_system.file_system_representation import FileSystemRepresentation

if TYPE_CHECKING:
    from posthog.models.team import Team


class ChatConversation(FileSystemSyncMixin, UUIDModel):
    class Meta:
        db_table = "posthog_chat_conversation"
        # No need for the unique constraint since we removed conversation_id
        pass

    person = models.ForeignKey(
        "posthog.Person",
        on_delete=models.CASCADE,
        related_name="chat_conversations",
        related_query_name="chat_conversation",
    )

    team = models.ForeignKey(
        "posthog.Team",
        on_delete=models.CASCADE,
        related_name="all_chat_conversations",
        related_query_name="all_chat_conversation",
    )

    title = models.CharField(max_length=400, blank=True, null=True)

    # Metadata
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    # Which page/url the conversation started on
    source_url = models.URLField(max_length=2000, blank=True, null=True)

    # Unread message count
    unread_count = models.PositiveIntegerField(default=0)

    @classmethod
    def get_file_system_unfiled(cls, team: "Team") -> QuerySet["ChatConversation"]:
        base_qs = cls.objects.filter(team=team)
        return cls._filter_unfiled_queryset(base_qs, team, type="chat_conversation", ref_field="id")

    def get_file_system_representation(self) -> FileSystemRepresentation:
        return FileSystemRepresentation(
            base_folder=self._create_in_folder or "Unfiled/Chats",
            type="chat_conversation",
            ref=str(self.pk),
            name=self.title
            or f"Chat with {self.person.properties.get('name') or self.person.properties.get('email') or f'Person {self.person.id}'}",
            href=f"/chat/{self.pk}",
            meta={
                "created_at": str(self.created_at),
                "person_id": str(self.person.id),
            },
            should_delete=False,
        )


class ChatMessage(UUIDModel):
    conversation = models.ForeignKey(
        ChatConversation,
        on_delete=models.CASCADE,
        related_name="messages",
        related_query_name="message",
    )

    content = models.TextField()

    # Metadata
    created_at = models.DateTimeField(auto_now_add=True)

    # Message has been read by the other party
    read = models.BooleanField(default=False)

    is_assistant = models.BooleanField(default=False, db_column="is_assistant")

    class Meta:
        ordering = ["created_at"]
        db_table = "posthog_chat_message"
