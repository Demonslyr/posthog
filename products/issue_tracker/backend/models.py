from django.db import models
from django.utils import timezone
import uuid


class Issue(models.Model):
    class Status(models.TextChoices):
        BACKLOG = "backlog", "Backlog"
        TODO = "todo", "To Do"
        IN_PROGRESS = "in_progress", "In Progress"
        TESTING = "testing", "Testing"
        DONE = "done", "Done"

    class OriginProduct(models.TextChoices):
        ERROR_TRACKING = "error_tracking", "Error Tracking"
        EVAL_CLUSTERS = "eval_clusters", "Eval Clusters"
        USER_CREATED = "user_created", "User Created"
        SUPPORT_QUEUE = "support_queue", "Support Queue"

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    title = models.CharField(max_length=255)
    description = models.TextField()
    status = models.CharField(max_length=20, choices=Status.choices, default=Status.BACKLOG)
    origin_product = models.CharField(max_length=20, choices=OriginProduct.choices)
    position = models.IntegerField(default=0)
    created_at = models.DateTimeField(default=timezone.now)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        db_table = "posthog_issue"
        managed = True
        ordering = ["position"]

    def __str__(self):
        return f"{self.title} ({self.get_status_display()})"
