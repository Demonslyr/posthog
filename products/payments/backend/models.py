from typing import TYPE_CHECKING
from django.db import models
from posthog.models.utils import UUIDModel, sane_repr

if TYPE_CHECKING:
    pass


class PaymentProduct(UUIDModel):
    class Meta:
        db_table = "posthog_payments_products"

    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    stripe_product_id = models.CharField(max_length=100)
    stripe_pricing_id = models.CharField(max_length=100)
    name = models.CharField(max_length=200)
    description = models.TextField(blank=True)
    price = models.DecimalField(max_digits=10, decimal_places=2)
    currency = models.CharField(max_length=3)  # e.g. USD, EUR
    date_created = models.DateTimeField(auto_now_add=True)
    date_updated = models.DateTimeField(auto_now=True)
    deleted = models.BooleanField(default=False)
    is_active = models.BooleanField(default=True)

    def __str__(self) -> str:
        return self.name

    __repr__ = sane_repr("id", "name", "team_id", "stripe_product_id")


class PaymentTransaction(UUIDModel):
    class Meta:
        db_table = "posthog_payments_transactions"

    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    product = models.ForeignKey(PaymentProduct, on_delete=models.PROTECT)
    payload = models.JSONField()
    status = models.CharField(max_length=50)
    date_created = models.DateTimeField(auto_now_add=True)

    def __str__(self) -> str:
        return f"{self.team_id} - {self.product.name} - {self.status}"

    __repr__ = sane_repr("id", "team_id", "product_id", "status")
