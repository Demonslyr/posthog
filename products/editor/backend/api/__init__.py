from .codebase_sync import CodebaseSyncViewset
from .max_tools import MaxToolsViewSet
from .proxy import SUPPORTED_MODELS_WITH_THINKING, LLMProxyViewSet

__all__ = ["CodebaseSyncViewset", "LLMProxyViewSet", "MaxToolsViewSet", "SUPPORTED_MODELS_WITH_THINKING"]
