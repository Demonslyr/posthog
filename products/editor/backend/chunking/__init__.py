from typing import cast

from .code import chunk_code as _chunk_code
from .exceptions import UnsupportedLanguage
from .parser import ProgrammingLanguage, guess_language
from .text import chunk_text as _chunk_text


def chunk_text(
    content: str, language: ProgrammingLanguage | None = None, chunk_size: int = 300, chunk_overlap: float = 0.2
):
    if language is not None:
        return _chunk_code(language, content, chunk_size, chunk_overlap)
    return _chunk_text(content, chunk_size, chunk_overlap)


__all__ = ["ProgrammingLanguage", "chunk_text", "guess_language"]
