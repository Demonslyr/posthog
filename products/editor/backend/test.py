import json
from datetime import datetime

from pydantic import BaseModel, Field

from posthog.clickhouse.client.execute import sync_execute
from posthog.test.base import BaseTest
from products.editor.backend.models.codebase import Codebase
from products.editor.backend.services.codebase_sync import SerializedArtifact


class CatalogEntry(BaseModel):
    team_id: int | None = Field(default=None)
    user_id: int | None = Field(default=None)
    codebase_id: str | None = Field(default=None)
    artifact_id: str
    parent_artifact_id: str | None = Field(default=None)
    branch: str | None = Field(default=None)
    type: str = Field(default="dir")
    sign: int = Field(default=1)
    timestamp: datetime | None = Field(default=None)


class EditorTestQueryHelpersMixin(BaseTest):
    codebase: Codebase
    stable_user_id: int = 99999

    def _create_artifacts(self, tree: list[SerializedArtifact]):
        query = "INSERT INTO codebase_embeddings (team_id, user_id, codebase_id, artifact_id, chunk_id, vector, properties, is_deleted) VALUES "
        rows: list[str] = []

        args = {
            "team_id": self.team.id,
            "user_id": self.user.id,
            "codebase_id": self.codebase.id,
            "chunk_id": 0,
            "vector": [0.5, 0.5],
            "properties": json.dumps(
                {
                    "lineStart": 0,
                    "lineEnd": 30,
                    "path": "obfuscated_path",
                }
            ),
            "is_deleted": 0,
        }

        for idx, node in enumerate(tree):
            if node["type"] == "file":
                args.update(
                    {
                        f"artifact_id_{idx}": node["id"],
                    }
                )
                rows.append(
                    f"(%(team_id)s, %(user_id)s, %(codebase_id)s, %(artifact_id_{idx})s, %(chunk_id)s, %(vector)s, %(properties)s, %(is_deleted)s)"
                )

        sync_execute(query + ", ".join(rows), args, team_id=self.team.id)

    def _create_codebase_catalog(self, tree: list[CatalogEntry]):
        query = "INSERT INTO codebase_catalog (team_id, user_id, codebase_id, artifact_id, parent_artifact_id, branch, type, timestamp, sign) VALUES "
        rows: list[str] = []
        args = {}

        for idx, node in enumerate(tree):
            args.update(
                {
                    f"team_id_{idx}": node.team_id or self.team.id,
                    f"user_id_{idx}": node.user_id or self.stable_user_id,
                    f"codebase_id_{idx}": node.codebase_id or str(self.codebase.id),
                    f"artifact_id_{idx}": node.artifact_id,
                    f"parent_artifact_id_{idx}": node.parent_artifact_id,
                    f"branch_{idx}": node.branch or "",
                    f"type_{idx}": node.type,
                    f"timestamp_{idx}": node.timestamp,
                    f"sign_{idx}": node.sign,
                }
            )
            rows.append(
                f"(%(team_id_{idx})s, %(user_id_{idx})s, %(codebase_id_{idx})s, %(artifact_id_{idx})s, %(parent_artifact_id_{idx})s, %(branch_{idx})s, %(type_{idx})s, %(timestamp_{idx})s, %(sign_{idx})s)"
            )

        sync_execute(query + ", ".join(rows), args, team_id=self.team.id)
