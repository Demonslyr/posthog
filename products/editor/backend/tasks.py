import json

from celery import shared_task
from openai import OpenAI
from structlog import get_logger

from posthog.clickhouse.client import sync_execute
from posthog.clickhouse.query_tagging import tag_queries
from products.editor.backend.chunking import chunk_text
from products.editor.backend.chunking.parser import ProgrammingLanguage
from products.editor.backend.chunking.types import Chunk

logger = get_logger(__name__)

EmbeddingResult = list[tuple[Chunk, list[float]]]


def chunk_and_embed(
    file_content: str, language: ProgrammingLanguage | None = None, batch_size: int = 2048
) -> EmbeddingResult:
    chunks = chunk_text(file_content, language=language)
    client = OpenAI()
    embeddings: list[tuple[Chunk, list[float]]] = []
    for batch_start in range(0, len(chunks), batch_size):
        chunk_slice = chunks[batch_start : batch_start + batch_size]
        response = client.embeddings.create(
            input=[chunk.text for chunk in chunk_slice],
            model="text-embedding-3-small",
        )
        for chunk, embedding_res in zip(chunk_slice, response.data):
            embeddings.append((chunk, embedding_res.embedding))

    return embeddings


def insert_embeddings(
    team_id: int,
    user_id: int,
    codebase_id: int,
    artifact_id: str,
    file_path: str,
    embeddings: EmbeddingResult,
):
    query = "INSERT INTO codebase_embeddings (team_id, user_id, codebase_id, artifact_id, chunk_id, vector, properties) VALUES "
    args = {
        "team_id": team_id,
        "user_id": user_id,
        "codebase_id": codebase_id,
        "artifact_id": artifact_id,
    }
    rows: list[str] = []

    for idx, (chunk, embedding) in enumerate(embeddings):
        args.update(
            {
                f"chunk_id_{idx}": str(idx),
                f"vector_{idx}": embedding,
                f"properties_{idx}": json.dumps(
                    {
                        "path": file_path,
                        "line_start": chunk.line_start,
                        "line_end": chunk.line_end,
                    }
                ),
            }
        )
        rows.append(
            f"(%(team_id)s, %(user_id)s, %(codebase_id)s, %(artifact_id)s, %(chunk_id_{idx})s, %(vector_{idx})s, %(properties_{idx})s)"
        )

    tag_queries(team_id=team_id, user_id=user_id, query_type="insert_embeddings", product="vsc-editor")
    sync_execute(query + ", ".join(rows), args, team_id=team_id)


@shared_task(ignore_result=True, max_retries=5, retry_backoff=True, expires=60 * 10)
def embed_file(
    team_id: int,
    user_id: int,
    codebase_id: int,
    artifact_id: str,
    file_path: str,
    file_content: str,
    language: ProgrammingLanguage | None = None,
):
    try:
        logger.info(
            "Embedding file",
            team_id=team_id,
            user_id=user_id,
            codebase_id=codebase_id,
            artifact_id=artifact_id,
        )
        embeddings = chunk_and_embed(file_content, language)
        insert_embeddings(team_id, user_id, codebase_id, artifact_id, file_path, embeddings)
    except Exception as e:
        logger.exception(
            "Error embedding file",
            team_id=team_id,
            user_id=user_id,
            codebase_id=codebase_id,
            artifact_id=artifact_id,
            error=type(e).__name__,
        )
        raise
