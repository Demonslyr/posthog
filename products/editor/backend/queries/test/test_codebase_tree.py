from posthog.hogql.constants import LimitContext
from posthog.schema import CodebaseTreeQuery, CodebaseTreeResponseItem
from posthog.test.base import BaseTest, ClickhouseTestMixin, snapshot_clickhouse_queries
from products.editor.backend.models.codebase import Codebase
from products.editor.backend.test import CatalogEntry, EditorTestQueryHelpersMixin

from ..codebase_tree import CodebaseTreeQueryRunner


class TestCodebaseTree(EditorTestQueryHelpersMixin, ClickhouseTestMixin, BaseTest):
    def setUp(self):
        super().setUp()
        self.codebase = Codebase.objects.create(team=self.team, user=self.user)

    @snapshot_clickhouse_queries
    def test_returns_codebase_tree(self):
        self._create_codebase_catalog(
            [
                CatalogEntry(artifact_id="root", parent_artifact_id=None, branch="main"),
                CatalogEntry(artifact_id="dir", parent_artifact_id="root", branch="main"),
                CatalogEntry(artifact_id="file", parent_artifact_id="dir", type="file", branch="main"),
            ]
        )

        response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id), branch="main"), self.team
        ).run()

        expected_result = [
            CodebaseTreeResponseItem(id="root", parentId=None, synced=True, type="dir"),
            CodebaseTreeResponseItem(id="dir", parentId="root", synced=True, type="dir"),
            CodebaseTreeResponseItem(id="file", parentId="dir", synced=False, type="file"),
        ]

        self.assertCountEqual(expected_result, response.results)

    @snapshot_clickhouse_queries
    def test_handles_empty_branch(self):
        """`branch` is a required string. If it's not provided or none, it must use an empty string."""
        self._create_codebase_catalog([CatalogEntry(artifact_id="root")])
        expected_result = [CodebaseTreeResponseItem(id="root", parentId=None, synced=True, type="dir")]

        response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id)), self.team
        ).run()
        self.assertCountEqual(expected_result, response.results)

        response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id), branch=""), self.team
        ).run()
        self.assertCountEqual(expected_result, response.results)

    @snapshot_clickhouse_queries
    def test_applies_higher_row_limit(self):
        """The query must apply a higher row limit than the default of 50k."""
        with self.capture_select_queries() as queries:
            CodebaseTreeQueryRunner(
                CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id)),
                self.team,
                limit_context=LimitContext.EDITOR,
            ).run()
            self.assertIn("LIMIT 1000000", queries[0])

    def test_synced_artifacts(self):
        """The response items must have synced=True if the artifact is synced."""
        self._create_codebase_catalog(
            [
                CatalogEntry(artifact_id="root"),
                CatalogEntry(artifact_id="file1", parent_artifact_id="root", type="file"),
                CatalogEntry(artifact_id="file2", parent_artifact_id="root", type="file"),
            ]
        )
        self._create_artifacts([{"id": "file1", "type": "file", "parent_id": "root"}])

        response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id)), self.team
        ).run()

        expected_result = [
            CodebaseTreeResponseItem(id="root", parentId=None, synced=True, type="dir"),
            CodebaseTreeResponseItem(id="file1", parentId="root", synced=True, type="file"),
            CodebaseTreeResponseItem(id="file2", parentId="root", synced=False, type="file"),
        ]
        self.assertCountEqual(expected_result, response.results)

    def test_multiple_users(self):
        """Test should return only artifacts for the specified user."""
        # First user (self.user) data
        self._create_codebase_catalog(
            [
                CatalogEntry(artifact_id="root", user_id=self.user.id),
                CatalogEntry(artifact_id="file1", parent_artifact_id="root", type="file", user_id=self.user.id),
            ]
        )

        # Second user data (static ID 999)
        other_user_id = 999
        self._create_codebase_catalog(
            [
                CatalogEntry(artifact_id="other-root", user_id=other_user_id),
                CatalogEntry(
                    artifact_id="other-file", parent_artifact_id="other-root", type="file", user_id=other_user_id
                ),
            ]
        )

        # Query for first user
        first_user_response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id)), self.team
        ).run()

        first_user_expected = [
            CodebaseTreeResponseItem(id="root", parentId=None, synced=True, type="dir"),
            CodebaseTreeResponseItem(id="file1", parentId="root", synced=False, type="file"),
        ]
        self.assertCountEqual(first_user_expected, first_user_response.results)

        # Query for second user
        second_user_response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=other_user_id, codebaseId=str(self.codebase.id)), self.team
        ).run()

        second_user_expected = [
            CodebaseTreeResponseItem(id="other-root", parentId=None, synced=True, type="dir"),
            CodebaseTreeResponseItem(id="other-file", parentId="other-root", synced=False, type="file"),
        ]
        self.assertCountEqual(second_user_expected, second_user_response.results)

    def test_collapsed_sign_rows(self):
        """Test that rows with sign=1 and sign=-1 are properly collapsed."""
        from datetime import datetime

        now = datetime.now()

        # Create row with sign=1
        self._create_codebase_catalog(
            [
                CatalogEntry(artifact_id="item1", timestamp=now, sign=1),
                # Create exact same row but with sign=-1, should collapse
                CatalogEntry(artifact_id="item1", timestamp=now, sign=-1),
                # This should appear in results
                CatalogEntry(artifact_id="item2", timestamp=now, sign=1),
            ]
        )

        response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id)), self.team
        ).run()

        # Only item2 should remain as item1 had sign 1 and -1 canceling each other
        expected_result = [
            CodebaseTreeResponseItem(id="item2", parentId=None, synced=True, type="dir"),
        ]
        self.assertCountEqual(expected_result, response.results)

    def test_timestamp_priority(self):
        """Test that the latest row with sign=1 is prioritized over later rows with sign=-1."""
        from datetime import datetime, timedelta

        now = datetime.now()
        earlier = now - timedelta(hours=1)
        later = now + timedelta(hours=1)

        self._create_codebase_catalog(
            [
                # Create initial entry
                CatalogEntry(artifact_id="file", parent_artifact_id="dir1", type="file", timestamp=earlier, sign=1),
                # Create an update with different parent
                CatalogEntry(artifact_id="file", parent_artifact_id="dir2", type="file", timestamp=now, sign=1),
                # Create a delete entry with even later timestamp - should NOT take precedence
                CatalogEntry(artifact_id="file", parent_artifact_id="dir2", type="file", timestamp=later, sign=-1),
                # Add the parent directories
                CatalogEntry(artifact_id="dir1", timestamp=earlier, sign=1),
                CatalogEntry(artifact_id="dir2", timestamp=earlier, sign=1),
            ]
        )

        response = CodebaseTreeQueryRunner(
            CodebaseTreeQuery(userId=self.user.id, codebaseId=str(self.codebase.id)), self.team
        ).run()

        # The file should still appear with parent dir2, even though a later sign=-1 row exists
        expected_result = [
            CodebaseTreeResponseItem(id="dir1", parentId=None, synced=True, type="dir"),
            CodebaseTreeResponseItem(id="dir2", parentId=None, synced=True, type="dir"),
            CodebaseTreeResponseItem(id="file", parentId="dir2", synced=False, type="file"),
        ]
        self.assertCountEqual(expected_result, response.results)
