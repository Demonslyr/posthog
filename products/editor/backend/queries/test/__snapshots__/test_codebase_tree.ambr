# serializer version: 1
# name: TestCodebaseTree.test_applies_higher_row_limit
  '''
  SELECT codebase_catalog.artifact_id AS artifact_id,
         argMaxIf(codebase_catalog.parent_artifact_id, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS parent_artifact_id,
         argMaxIf(codebase_catalog.type, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS artifact_type,
         or(ifNull(equals(artifact_type, 'dir'), 0), ifNull(notEquals(any(embeddings.synced_artifact_id), ''), 1)) AS synced
  FROM codebase_catalog
  LEFT JOIN
    (SELECT argMax(DISTINCT codebase_embeddings.artifact_id, toTimeZone(codebase_embeddings.timestamp, 'UTC')) AS synced_artifact_id
     FROM codebase_embeddings
     WHERE and(equals(codebase_embeddings.team_id, 99999), equals(codebase_embeddings.user_id, 308), equals(codebase_embeddings.codebase_id, '019634b6-b3dd-0000-684b-b69f8c084ced'))
     GROUP BY codebase_embeddings.artifact_id) AS embeddings ON equals(codebase_catalog.artifact_id, embeddings.synced_artifact_id)
  WHERE and(equals(codebase_catalog.team_id, 99999), equals(codebase_catalog.user_id, 308), equals(codebase_catalog.codebase_id, '019634b6-b3dd-0000-684b-b69f8c084ced'), equals(codebase_catalog.branch, ''))
  GROUP BY codebase_catalog.artifact_id
  HAVING ifNull(greater(sum(codebase_catalog.sign), 0), 0)
  LIMIT 1000000 SETTINGS readonly=2,
                         max_execution_time=60,
                         allow_experimental_object_type=1,
                         format_csv_allow_double_quotes=0,
                         max_ast_elements=4000000,
                         max_expanded_ast_elements=4000000,
                         max_bytes_before_external_group_by=0,
                         transform_null_in=1,
                         optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestCodebaseTree.test_handles_empty_branch
  '''
  SELECT codebase_catalog.artifact_id AS artifact_id,
         argMaxIf(codebase_catalog.parent_artifact_id, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS parent_artifact_id,
         argMaxIf(codebase_catalog.type, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS artifact_type,
         or(ifNull(equals(artifact_type, 'dir'), 0), ifNull(notEquals(any(embeddings.synced_artifact_id), ''), 1)) AS synced
  FROM codebase_catalog
  LEFT JOIN
    (SELECT argMax(DISTINCT codebase_embeddings.artifact_id, toTimeZone(codebase_embeddings.timestamp, 'UTC')) AS synced_artifact_id
     FROM codebase_embeddings
     WHERE and(equals(codebase_embeddings.team_id, 99999), equals(codebase_embeddings.user_id, 310), equals(codebase_embeddings.codebase_id, '019634b6-b6a0-0000-aae8-a9ce83b2fb71'))
     GROUP BY codebase_embeddings.artifact_id) AS embeddings ON equals(codebase_catalog.artifact_id, embeddings.synced_artifact_id)
  WHERE and(equals(codebase_catalog.team_id, 99999), equals(codebase_catalog.user_id, 310), equals(codebase_catalog.codebase_id, '019634b6-b6a0-0000-aae8-a9ce83b2fb71'), equals(codebase_catalog.branch, ''))
  GROUP BY codebase_catalog.artifact_id
  HAVING ifNull(greater(sum(codebase_catalog.sign), 0), 0)
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestCodebaseTree.test_handles_empty_branch.1
  '''
  SELECT codebase_catalog.artifact_id AS artifact_id,
         argMaxIf(codebase_catalog.parent_artifact_id, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS parent_artifact_id,
         argMaxIf(codebase_catalog.type, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS artifact_type,
         or(ifNull(equals(artifact_type, 'dir'), 0), ifNull(notEquals(any(embeddings.synced_artifact_id), ''), 1)) AS synced
  FROM codebase_catalog
  LEFT JOIN
    (SELECT argMax(DISTINCT codebase_embeddings.artifact_id, toTimeZone(codebase_embeddings.timestamp, 'UTC')) AS synced_artifact_id
     FROM codebase_embeddings
     WHERE and(equals(codebase_embeddings.team_id, 99999), equals(codebase_embeddings.user_id, 310), equals(codebase_embeddings.codebase_id, '019634b6-b6a0-0000-aae8-a9ce83b2fb71'))
     GROUP BY codebase_embeddings.artifact_id) AS embeddings ON equals(codebase_catalog.artifact_id, embeddings.synced_artifact_id)
  WHERE and(equals(codebase_catalog.team_id, 99999), equals(codebase_catalog.user_id, 310), equals(codebase_catalog.codebase_id, '019634b6-b6a0-0000-aae8-a9ce83b2fb71'), equals(codebase_catalog.branch, ''))
  GROUP BY codebase_catalog.artifact_id
  HAVING ifNull(greater(sum(codebase_catalog.sign), 0), 0)
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestCodebaseTree.test_returns_codebase_tree
  '''
  SELECT codebase_catalog.artifact_id AS artifact_id,
         argMaxIf(codebase_catalog.parent_artifact_id, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS parent_artifact_id,
         argMaxIf(codebase_catalog.type, toTimeZone(codebase_catalog.timestamp, 'UTC'), equals(codebase_catalog.sign, 1)) AS artifact_type,
         or(ifNull(equals(artifact_type, 'dir'), 0), ifNull(notEquals(any(embeddings.synced_artifact_id), ''), 1)) AS synced
  FROM codebase_catalog
  LEFT JOIN
    (SELECT argMax(DISTINCT codebase_embeddings.artifact_id, toTimeZone(codebase_embeddings.timestamp, 'UTC')) AS synced_artifact_id
     FROM codebase_embeddings
     WHERE and(equals(codebase_embeddings.team_id, 99999), equals(codebase_embeddings.user_id, 312), equals(codebase_embeddings.codebase_id, '019634b6-b924-0000-5f70-437f85cf9c25'))
     GROUP BY codebase_embeddings.artifact_id) AS embeddings ON equals(codebase_catalog.artifact_id, embeddings.synced_artifact_id)
  WHERE and(equals(codebase_catalog.team_id, 99999), equals(codebase_catalog.user_id, 312), equals(codebase_catalog.codebase_id, '019634b6-b924-0000-5f70-437f85cf9c25'), equals(codebase_catalog.branch, 'main'))
  GROUP BY codebase_catalog.artifact_id
  HAVING ifNull(greater(sum(codebase_catalog.sign), 0), 0)
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
